### PR TITLE
Refactor configuration to be standardized and in one place

### PR DESCRIPTION
**Pull request by bmedx**
Refactors configuration into an object, centralizing annotation parsing.

### CODE REVIEW

File: code_annotations/base.py [hunk 1]
```diff
@@ -10,115 +10,261 @@
...
+
+            elif not self._is_annotation_token(annotation):  # pragma: no cover
                 # If not we don't know what to do with it
                 raise TypeError(
-                    '{} is an unknown type. Annotations must be strings or list/tuples.'.format(annotation_or_group)
+                    '{} is an unknown type, must be strings or lists.'.format(annotation_token_or_group_name)
                 )
+            else:
+                self.annotation_tokens.append(annotation_token_or_group_name)
+                self.annotation_regexes.append(re.escape(annotation_token_or_group_name))
```
Comment (doctoryes) [inline code_annotations/base.py]: This logic is much easier to understand and maintain.

File: code_annotations/base.py [hunk 2]
```diff
@@ -262,25 +414,23 @@ def check_results(self, all_results):
...
-                    if token in self.groups:
-                        self.echo.echo_vv('Starting new group for "{}" line {}'.format(
-                            token, annotation['line_number'])
+                    if token in group_children:
+                        current_group = self._get_group_for_token(token)
+
+                        if current_group is None:  # pragma: no cover
+                            # If we get here there is a problem with check_results' group_children not matching up with
+                            # our config's groups. That puts us in an unknown state, so we should quit.
+                            raise Exception('group_children is out of sync with config.groups!')
```
Comment (doctoryes) [inline code_annotations/base.py]: Would it be worth calling out the value of `token` here in the exception?

File: tests/extensions/python_test_files/group_failures_3.pyt [hunk 3]
```diff
@@ -1,7 +0,0 @@
-"""
-Test group with first item out of order
```
Comment (pwnage101) [inline tests/extensions/python_test_files/group_failures_3.pyt]: Does this mean that group annotations can be out of order now?  Regardless, Shouldn't we retain this test case?

File: tests/helpers.py [hunk 4]
```diff
@@ -6,29 +6,57 @@
...
         - ".. pii_retirement::":
             choices: [retained, local_api, consumer_api, third_party]
-    nopii: ".. no_pii::"
-    ignored:
-        ".. ignored::":
-            choices: [irrelevant, terrible, silly-silly]
+extensions:
+    python:
+        - py
 """
```
Comment (pwnage101) [inline tests/helpers.py]: While you're in here, why not bake fake_safelist_path into this string right now during module setup?

File: code_annotations/base.py [hunk 5]
```diff
@@ -10,115 +10,261 @@
...
+        for annotation in group:
+            for annotation_token in annotation:
+                annotation_value = annotation[annotation_token]
+
+                # The annotation comment is a choice group
+                if self._is_choice_group(annotation_value):
+                    self._configure_choices(annotation_token, annotation_value)
+                # Otherwise it should be a text type, if not then error out
+                elif not self._is_annotation_token(annotation_value):
+                    raise TypeError('{} is an unknown type.'.format(annotation))
```
Comment (pwnage101) [inline code_annotations/base.py]: Maybe this should be more specific, like "is an unknown annotation type".

File: tests/extensions/python_test_files/group_failures_3.pyt (hunk 3 above)
Comment (bmedx) [reply tests/extensions/python_test_files/group_failures_3.pyt]: It does mean that, yep. The tests in `group_ordering_2.pyt` test this now, so it was just redundant.

**Pull request merged**